spec/index.md

```diff
@@ -5033,6 +5094,179 @@ S with
+      memory_usage_raw_module(A.wasm_module) +
+      memory_usage_canister_history(New_canister_history) +
+      memory_usage_snapshot(New_snapshot),
+    S.canister_subnet[A.canister_id].subnet_size,
+  )
+) ≥ 0
+
+```
+
+State after  
+
+```html
+
+S with
+    snapshots[A.canister_id] = New_snapshot
+    balances[A.canister_id] = New_balance
+    reserved_balances[A.canister_id] = New_reserved_balance
+    messages = Older_messages · Younger_messages ·
+      ResponseMessage {
+        origin = M.origin;
+        response = Reply (candid(SnapshotId));
+        refunded_cycles = M.transferred_cycles;
+      }
+
+```
+
+
+#### IC Management Canister: Load snapshots
+
+
+Only the controllers of the given canister can load a snapshot.
+
+```html
+
+S.messages = Older_messages · CallMessage M · Younger_messages
+(M.queue = Unordered) or (∀ msg ∈ Older_messages. msg.queue ≠ M.queue)
+M.callee = ic_principal
+M.method_name = 'load_snapshot'
+M.arg = candid(A)
+M.caller ∈ S.controllers[A.canister_id]
+
+New_balance = S.balances[A.canister_id] - Cycles_used
+liquid_balance(
+  New_balance,
+  S.reserved_balances[A.canister_id],
+  freezing_limit(
+    S.compute_allocation[A.canister_id],
+    S.memory_allocation[A.canister_id],
+    S.freezing_threshold[A.canister_id],
+    memory_usage_wasm_state(S.canisters[A.canister_id]) +
+      memory_usage_raw_module(A.wasm_module) +
+      memory_usage_canister_history(New_canister_history) +
+      memory_usage_snapshot(S.snapshots[A.canister_id]),
+    S.canister_subnet[A.canister_id].subnet_size,
+  )
+) ≥ 0
+
+```
+
+State after  
+
+```html
+
+S with
+    canisters[A.canister_id] = {
+      wasm_state = S.snapshots[A.canister_id].wasm_state;
+      raw_module = S.snapshots[A.canister_id].raw_module;
+      module = S.snapshots[A.canister_id].wasm_module;
+      chunk_store = S.snapshots[A.canister_id].chunk_store;
+      public_custom_sections = parse_public_custom_sections(S.snapshots[A.canister_id].wasm_module);
+      private_custom_sections = parse_private_custom_sections(S.snapshots[A.canister_id].wasm_module);
+    }
+    balances[A.canister_id] = New_balance
+    messages = Older_messages · Younger_messages ·
+      ResponseMessage {
+        origin = M.origin;
+        response = Reply (candid());
+        refunded_cycles = M.transferred_cycles;
+      }
+
+```
+
+#### IC Management Canister: List snapshots
+
+Only the controllers of the given canister can get a list of the existing snapshots.
+
+```html
+
+S.messages = Older_messages · CallMessage M · Younger_messages
+(M.queue = Unordered) or (∀ msg ∈ Older_messages. msg.queue ≠ M.queue)
+M.callee = ic_principal
+M.method_name = 'list_snapshots'
+M.arg = candid(A)
+M.caller ∈ S.controllers[A.canister_id]
+
+```
+
+State after
+
+```html
+
+S with
+    messages = Older_messages · Younger_messages ·
+      ResponseMessage {
+        origin = M.origin
+        response = Reply( candid(dom(S.snapshots[A.canister_id])))
```

Comment:
Using `dom(...)` wouldn't work here since you define `snapshots` to be a single snapshot (not a map of snapshot IDs to their corresponding snapshots).